📖 iImprove testing guidelines (#6112)
kubernetes-sigs/cluster-api: docs/book/src/developer/testing.md

be fast and great for getting the first signal on the current implementation, but unit tests have the risk of
allowing integration bugs to slip through.

> In Cluster API most of the unit tests are developed using [go test], [gomega] and the [fakeclient]; however using
> [fakeclient] is not suitable for all the use cases due to some limitation derived by how it is implemented,
> so in some cases contributors will be required to use [envtest]. See the [quick reference](#quick-reference) below for more details.

Reviewer: ```suggestion
In Cluster API most of the unit tests are developed using [go test], [gomega] and the [fakeclient]; however using
[fakeclient] is not suitable for all the use cases due to some limitations in how it is implemented. In some cases contributors will be required to use [envtest]. See the [quick reference](#quick-reference) below for more details.
```

### Mocking external APIs

> In some cases when writing tests it is required to mock external API, e.g. etcd client API or the AWS sdk API.

Reviewer: ```suggestion
In some cases when writing tests it is required to mock external API, e.g. etcd client API or the AWS SDK API.
```

This problem is usually well scoped in core Cluster API, and in most cases it is already solved by using fake
implementations of the target API to be injected during tests.

Instead, mocking is much more relevant for infrastructure providers; in order to address the issue

> some providers can use simulators reproducing the behaviour of a real infrastructure providers (e.g CAPV);

Reviewer 2: Simulators have advantages, but also come with a host of their own problems. I'm not sure I'd explicitly favour one over the other, certainly not in the context of unit tests.

> if this is not possible, a viable solution is to use mocks (e.g CAPA).

Reviewer 2: CAPO are also using mocks extensively, btw.

> if this is not possible, a viable solution is to use mocks (e.g CAPA).

Author: The goal of this paragraph is to create awareness on the topic and give a pointer to where this is implemented outside of the CAPI repo, but I'm personally -1 to add a link to unit test code in another repository given that this will be brittle.

### Generic providers

> While testing core Cluster API contributors should ensure that the code works with any providers, and thus it is required
> to not use any specific provider implementation. Instead, the so-called generic providers should be used because they
> implement the plain Cluster API contract, thus preventing developers to make assumptions we cannot rely on.

Reviewer: ```suggestion
When writing tests core Cluster API contributors should ensure that the code works with any providers, and thus it is required
to not use any specific provider implementation. Instead, the so-called generic providers e.g. "GenericInfrastructureCluster" should be used because they
implement the plain Cluster API contract. This prevents tests from relying on assumptions that may not hold true in all cases.
```

Possibly add a link to the generic providers in the code? Or at least the name of the type so they can be searched.

Author: I'm a little bit concerned about this because what we have is really fragmented. Even if not ideal, I prefer contributors to ask for guidance on this point so we can weigh pros and cons case by case; also, using the v1.1 timeframe as a benchmark, changes of this type are not really frequent.

Reviewer: I think this is fine, given that a link might be more confusing. Adding the name of the example type, i.e. GenericInfrastructureCluster,  lets people search for it themselves in the codebase to see the implementation

> Please note that in the long tern we would like to improve current implementation of generic providers, centralizing

Reviewer 3: typo: tern -> term

the existing set of utilities scattered across the codebase, but while details of this work will be defined do not
hesitate to reach out to reviewers and maintainers for guidance.

## Integration tests

Integration tests are focused on testing the behavior of an entire controller or the interactions between two or
more Cluster API controllers.

> In Cluster API, integration test are based on [envtest] and one or more controllers configured to run against

Reviewer: ```suggestion
In Cluster API, integration tests are based on [envtest] and one or more controllers configured to run against
```

- it does not properly handle a set of fields which are common in the Kubernetes API objects (and Cluster API objects as well)
  like e.g. `creationTimestamp`, `resourceVersion`, `generation`, `uid`

> - API calls does not execute defaulting or validation webhooks, so there are no enforced guarantees about the semantic accuracy

Reviewer: ```suggestion
- API calls do not execute defaulting or validation webhooks, so there are no enforced guarantees about the semantic accuracy
```
